cdap-client/src/main/java/co/cask/cdap/client/MetadataClient.java

```diff
@@ -0,0 +1,473 @@
+   * @param properties tags to be added
+   */
+  public void addProperties(Id.Application appId, Map<String, String> properties)
+    throws IOException, UnauthorizedException, NotFoundException, BadRequestException {
+    String path = String.format("apps/%s/metadata/properties", appId.getId());
+    addProperties(appId, path, properties);
+  }
+
+  /**
+   * Adds properties to a dataset.
+   *
+   * @param datasetInstance dataset to add tags to
+   * @param properties tags to be added
+   */
+  public void addProperties(Id.DatasetInstance datasetInstance, Map<String, String> properties)
+    throws IOException, UnauthorizedException, NotFoundException, BadRequestException {
+    String path = String.format("datasets/%s/metadata/properties", datasetInstance.getId());
+    addProperties(datasetInstance, path, properties);
+  }
+
+  /**
+   * Adds properties to a stream.
+   *
+   * @param streamId stream to add tags to
+   * @param properties tags to be added
+   */
+  public void addProperties(Id.Stream streamId, Map<String, String> properties)
+    throws IOException, UnauthorizedException, NotFoundException, BadRequestException {
+    String path = String.format("streams/%s/metadata/properties", streamId.getId());
+    addProperties(streamId, path, properties);
+  }
+
+  /**
+   * Adds properties to a program.
+   *
+   * @param programId program to add tags to
+   * @param properties tags to be added
+   */
+  public void addProperties(Id.Program programId, Map<String, String> properties)
+    throws IOException, UnauthorizedException, NotFoundException, BadRequestException {
+    String path = String.format("apps/%s/%s/%s/metadata/properties",
+                                programId.getApplicationId(), programId.getType().getCategoryName(), programId.getId());
+    addProperties(programId, path, properties);
+  }
+
+  private void addProperties(Id.NamespacedId namespacedId, String path, Map<String, String> properties)
+    throws IOException, UnauthorizedException, NotFoundException, BadRequestException {
+    URL url = config.resolveNamespacedURLV3(namespacedId.getNamespace(), path);
+    HttpResponse response =
+      restClient.execute(HttpRequest.post(url).withBody(GSON.toJson(properties)).build(), config.getAccessToken(),
+                         HttpURLConnection.HTTP_BAD_REQUEST, HttpURLConnection.HTTP_NOT_FOUND);
+    if (response.getResponseCode() == HttpURLConnection.HTTP_BAD_REQUEST) {
+      throw new BadRequestException(response.getResponseBodyAsString());
+    }
+    if (response.getResponseCode() == HttpURLConnection.HTTP_NOT_FOUND) {
+      throw new NotFoundException(namespacedId);
+    }
+  }
+
+  /**
+   * Removes properties from an application.
+   *
+   * @param appId app to remove properties from
+   * @param propertyToRemove property to be removed, or {@code null} to remove all properties for the entity
+   */
+  public void removeProperties(Id.Application appId, @Nullable String propertyToRemove)
```

Comment:
@anwar6953, since this is a public client, it would be better to add methods like ``removeProperties(Id.Application appId)`` that just delegate to ``removeProperties(appId, null)``, so the API is simpler for users.